proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy

```diff
@@ -590,32 +781,44 @@ lemma perform_invocation_valid_cur_vcpu[wp]:
   done
 
 crunch reply_from_kernel, receive_signal
-  for valid_cur_vcpu[wp]: valid_cur_vcpu
-  (wp: valid_cur_vcpu_lift_weak)
+  for cur_domain[wp]: "\<lambda>s. P (cur_domain s)"
+  and etcbs_of[wp]: "\<lambda>s. P (etcbs_of s)"
+  and valid_cur_vcpu[wp]: valid_cur_vcpu
+  (wp: valid_cur_vcpu_lift_weak in_cur_domain_lift_weak simp: crunch_simps)
+
+lemma ct_in_cur_domain_active_resume_cur_thread:
+  "\<lbrakk>ct_in_cur_domain s; ct_active s; valid_idle s; scheduler_action s = resume_cur_thread\<rbrakk>
+   \<Longrightarrow> in_cur_domain (cur_thread s) s"
+  by (clarsimp simp: ct_in_cur_domain_def ct_in_state_def dest!: st_tcb_at_idle_thread)
 
 lemma handle_invocation_valid_cur_vcpu[wp]:
-  "\<lbrace>valid_cur_vcpu and invs and ct_active\<rbrace> handle_invocation calling blocking \<lbrace>\<lambda>_. valid_cur_vcpu\<rbrace>"
+  "\<lbrace>valid_cur_vcpu and einvs and ct_active and (\<lambda>s. scheduler_action s = resume_cur_thread)\<rbrace>
+   handle_invocation calling blocking
+   \<lbrace>\<lambda>_. valid_cur_vcpu\<rbrace>"
   unfolding handle_invocation_def
   by (wp syscall_valid set_thread_state_ct_st
-      | simp add: split_def | wpc
+      | simp add: split_def | wpc | wps
       | wp (once) hoare_drop_imps)+
-     (auto simp: ct_in_state_def elim: st_tcb_ex_cap)
+     (auto simp: ct_in_state_def valid_sched_def elim: st_tcb_ex_cap
+         intro!: ct_in_cur_domain_active_resume_cur_thread)
 
 lemma handle_recv_valid_cur_vcpu[wp]:
   "handle_recv is_blocking \<lbrace>valid_cur_vcpu\<rbrace>"
   unfolding handle_recv_def Let_def ep_ntfn_cap_case_helper delete_caller_cap_def
   by (wpsimp wp: hoare_drop_imps)
 
 lemma handle_event_valid_cur_vcpu:
-  "\<lbrace>valid_cur_vcpu and invs and (\<lambda>s. e \<noteq> Interrupt \<longrightarrow> ct_active s)\<rbrace>
+  "\<lbrace>valid_cur_vcpu and einvs and (\<lambda>s. e \<noteq> Interrupt \<longrightarrow> ct_active s)
```

Comment:
I know you had a bit of a battle with handle_event... can you say something (here, not as a comment) as to how we can use einvs here, or why, or how you came up with this schema?

Reply:
This isn't actually the scene of that specific battle, that's in `Syscall_AC`. I'm using `einvs` here for the existing `ct_in_cur_domain` invariant, which says that if the scheduler action is `resume_cur_thread` then the current thread is either the idle thread or in the current domain. I use that to resolve the various `in_cur_domain (cur_thread s) s` preconditions once I'm able to.